### ISSUE TITLE
Allow controllers to be either callable or constructible

### ISSUE THREAD

**Issue by isiahmeadows** [lhorie/mithril.js#994]
*(Branched off of [this comment of mine](https://github.com/lhorie/mithril.js/issues/914#issuecomment-198594199))*

**TL;DR:** Controllers should be *anything* executable: arrow functions, object methods, classes, etc.

If you have any ideas/opinions, good or bad, please share them! :smile: 

**Edit:** Pulled the second half out to #995.

---

This would make both arrow functions and ES6 classes valid controllers. The problem at hand is that controllers, due to internal design, can only be ES5 functions, since they are both called and constructed by Mithril. If you have ES6 support enabled in your browser, you'll notice that in the case of child components, they're [called](http://codepen.io/isiahmeadows/pen/pyRMJK), not [constructed](http://codepen.io/isiahmeadows/pen/eZgwqP). This is inconsistent with the documentation, which says [class constructors should be valid controllers](http://mithril.js.org/mithril.component.html#controller-as-a-class-constructor). 

This would require wrapping all controller constructions into this function below:

```js
function init(controller, args) {
  if ("prototype" in f) {
    return new (f.bind.apply(f, [undefined].concat(args)))()
  } else {
    var inst = {}
    var ret = f.apply(inst, args)
    return ret !== undefined ? ret : inst
  }
}
```

@barneycarroll suggested [creating a new branch](https://github.com/lhorie/mithril.js/issues/914#issuecomment-198718966) for developing this and #995 further and see where that takes us. Also, feel free to discuss any ideas you have about this here. :smile: 

---

/cc @lhorie @tivac @barneycarroll @pygy @ArthurClemens @StephanHoyer @mindeavor @ciscoheat

**Comment by lhorie** [comment]
My 2 cents: 1) and 2) seem totally unrelated. Is there a reason why you're putting both together?

> for mithril-objectify users, it's not possible to convert m(component, arg1, arg2) into something meaningful

With regards to the breaking change in 2), I don't actually mind that it is a breaking change, but I'm thinking it might be a good idea to log a warning for at least a version if the diff engine encounters that usage pattern.

**Comment by isiahmeadows** [comment]
@lhorie Ok. I'll edit the shit out of this bug to pull 2 out into its own thing. You have a good point, and I almost did that myself beforehand.

**Comment by isiahmeadows** [comment]
@lhorie #995

**Comment by tivac** [comment]
I want to see benchmarks, we've only seen a single issue on this so far.

**Comment by isiahmeadows** [comment]
@tivac I doubt it would cause much of a perf hit after the initial load or a full route change in practice, because controllers aren't frequently instantiated. This issue is more about ergonomics, and should hardly affect performance in practice (unless you're frequently making lots of top-level vdom modifications, like in route changes).

**Comment by tivac** [comment]
This seems like such an edge-case thing, especially since it can **only** be hit if you're running raw ES2015. Doesn't really seem worth a perf hit to me, which is why I'd like to see the work put in to justify that it isn't a big deal.

**Comment by isiahmeadows** [comment]
@tivac FWIW, Mithril already doesn't consistently call vs construct functions. See [this pen for classes not working](http://codepen.io/isiahmeadows/pen/eZgwqP) and [this pen for callables working](http://codepen.io/isiahmeadows/pen/pyRMJK) in the first level down, but [this pen for classes working](http://codepen.io/isiahmeadows/pen/ONpqKX) and [this pen for callables not working](http://codepen.io/isiahmeadows/pen/LNWaKv) at the top level. Note that your browsers need ES6 support to run these.

That's my other concern. My idea is to allow both, but even picking one is better than nothing. And performance-wise, it's probably going to spend more time actually running the controller function/class in practice than trying to construct it with that `init` helper. This would be almost instant 90% of the time, anyways, and because this isn't a frequent occurrence (on a given component, `view` might be called 1000 times a second, while `controller` might be called once every 5 seconds most of the time), I'm not worried after the first render of a route.

Another thing is that `bind` is unavoidable for ES6 constructors unless you're willing to use spread syntax.

**Comment by isiahmeadows** [comment]
And if you're just writing a Chrome or Firefox app, or if you plan on using Mithril 2 years down the road when ES6 support is ubiquitous, this is relevant.

**Comment by aliem** [comment]
my 0.2 cent:
I think controllers should be simple functions to prepare data. A function without a prototype should be enough, in the end you only need a simple object for the view, not a _throw away_ instance of a function.

Anyway the current behavior is very confusing while using a recent browser and mixed syntax (or simply ES6/2014).

The arrow functions used as controllers should work.

So I think the nice `this` convenience could be removed in favour of the more verbose `return` object pattern. It would not create more issues in the future and would be far more simple to reason about it.

```js
const component = {
  controller: () => ({
    data: 'Hello World',
  }),
  view(ctl) {
    return m('div', ctl.data);
  }
}
```

```js
const component = {
  controller() {
    return {
      data: 'Hello World',
    };
  },
  view(ctl) {
    return m('div', ctl.data);
  }
}
```

**Comment by isiahmeadows** [comment]
To be perfectly honest, I'm leaning towards the purely-callable controllers myself (still having a `this`). It's just as easy to implement, and it still allows arrow functions and ES6 methods.

The one concern I have is that TypeScript users might have a few issues migrating, but it should be straightforward to solve normally.

This was in part to invoke discussion on the topic, because I've been seeking an answer to some degree for a few months, now. The current functionality has been broken for that long. And since there are class die-hards as well as callable die-hards (@barneycarroll :wink:), I threw this out to generate some discussion on this.

Now, always returning an object would be impractical, because too many people rely on `this`. It might make sense if you use arrow functions, but not if you use `this`.

**Comment by barneycarroll** [comment]
Not sure if it makes anyone more comfortable, but we could forcibly apply constructor return logic as follows:

```javascript
function init(f, args) {
  if ("prototype" in f) {
    return new (f.bind.apply(f, [undefined].concat(args)))()
  } else {
    var inst = {}
    var ret = f.apply(inst, args)
    return ret instanceof Object ? ret : inst // [1]
  }
}
```

I'm at work at the moment (where jsbin is blocked by the corporate proxy), but I'm fairly sure this would lead to universally predictable outcomes for controller outputs regardless of function subtype.

**Comment by isiahmeadows** [comment]
@barneycarroll [Not quite](http://www.ecma-international.org/ecma-262/6.0/#sec-ecmascript-function-objects-construct-argumentslist-newtarget). Try `typeof ret === "object" && ret !== null` instead of `ret instanceof Object`. (Mine isn't correct, either. I misremembered.)

**Comment by barneycarroll** [comment]
I keep on forgetting `instanceof` isn't fit for purpose. But. Javascript can produce entities that pass `typeof ret === "object" && ret === null`?!

**Comment by tivac** [comment]
@barneycarroll yes.

```
> typeof null
> "object"
```

**Comment by isiahmeadows** [comment]
@barneycarroll Try `new function () { return function f() {} }`. That actually returns the inner `f`, since according to spec, it's just an object with a `[[Call]]` method.

**Comment by isiahmeadows** [comment]
That's the other difference.

**Comment by isiahmeadows** [comment]
@tivac Actually, the perf difference for that `init` function will be significant, but can be alleviated for most cases with this helper (which tries to avoid the otherwise expensive `bind` call):

```js
function constructWithArgs(C, args) {
    // Try to fast-path with a direct call first.
    switch (args.length) {
    case 0: return new C()
    case 1: return new C(args[0])
    case 2: return new C(args[0], args[1])
    case 3: return new C(args[0], args[1], args[2])
    case 4: return new C(args[0], args[1], args[2], args[3])
    default: return new (C.bind.apply(C, [undefined].concat(args)))()
    }
}
```

**Comment by lhorie** [comment]
since controllers are gone in rewrite, I'm closing this
